### PR TITLE
[INFRA-1690] One image, change jdk at start

### PR DESCRIPTION
**Pull request by MRamonLeon**
See [[INFRA-1690] DevTools: Add support of Specifying Java Level for Test in runATH()](https://issues.jenkins-ci.org/browse/INFRA-1690)

**TL;DR**
This approach build the image with both jdks and change the java in use when starting the container.

Currently the docker image generated is failing because the Dockerfile needs the `java_version` argument to build the image correctly. In addition, we need a way to run the `runATH` step in the _pipeline-library_ using both JDKs without building the ATH on the fly.

To do that we run a script during the _run_ of the container setting all _java programs_ to the right version. In addition, print a message in the container console to help starting with the tests using the right commands depending on the JDK in use. Example with Java11:
```
Running on...
openjdk version "11.0.1" 2018-10-16
OpenJDK Runtime Environment (build 11.0.1+13-Debian-2)
OpenJDK 64-Bit Server VM (build 11.0.1+13-Debian-2, mixed mode, sharing)

javac 11.0.1

Start running tests with...
eval $(./vnc.sh)

env JENKINS_OPTS="--enable-future-java" ./run.sh firefox latest -DforkCount=1 -Dmaven.test.failure.ignore=true -B -Dtest=...
```

The instructions don't change. As usual:
* Build an get into the container:
   * Java 8: `./ath-container.sh`
   * Java 11: `java_version=11 ./ath-container.sh`

* Set vncserver:
   * `eval $(./vnc.sh)`

* Run tests:
   * Java 8: `./run.sh ...`
   * Java 11: `JENKINS_OPTS="--enable-future-java" ./run.sh ...`

### CODE REVIEW

**Comment by raul-arabaolaza** [comment]
I greatly like the idea of making everything work by default in the same way than before the java 11 changes. So with this changes anything that was working before java 11 related changes is going to work. In other words any job constructing the docker image is going to be back to normal and any job running the ath using the docker image is going to be back to normal.

+500E108564734 on my side

File: set-java.sh [hunk 1]
```diff
@@ -0,0 +1,123 @@
+#!/bin/bash
+
+# The selection used by update-alternatives for each java version
+if [ "$1" == "11" ]; then
+    # Java 11
+    selection="1"
+    runcommand="env JENKINS_OPTS=\"--enable-future-java\" ./run.sh firefox latest -DforkCount=1 -Dmaven.test.failure.ignore=true -B -Dtest=..."
```
Comment (oleg-nenashev) [inline set-java.sh]: should not it also pass Java modules here? or is it handled elsewhere?

Comment (olivergondza) [reply set-java.sh]: Have not turned out to be needed at all.

File: ath-container.sh [hunk 2]
```diff
@@ -8,8 +8,8 @@ gid=$(id -g)
 tag="jenkins/ath"
 java_version="${java_version:-8}"
 
-docker build --build-arg=uid="$uid" --build-arg=gid="$gid" --build-arg=java_version="$java_version" src/main/resources/ath-container -t "$tag"
+docker build --build-arg=uid="$uid" --build-arg=gid="$gid" src/main/resources/ath-container -t "$tag"
 
 run_opts="--interactive --tty --rm --publish-all --user ath-user --workdir /home/ath-user/ath-sources"
 run_drive_mapping="-v /var/run/docker.sock:/var/run/docker.sock -v $(pwd):/home/ath-user/ath-sources -v ${HOME}/.m2/repository:/home/ath-user/.m2/repository"
-docker run --shm-size 2g $run_opts $run_drive_mapping $tag /bin/bash
+docker run --shm-size 2g $run_opts $run_drive_mapping $tag /bin/bash -c "./set-java.sh $java_version; bash"
```
Comment (olivergondza) [inline ath-container.sh]: -1 on the approach here. Note the container can / is supposed to be used interactively - not just to invoke the testsuite. IMO we have to separate that, providing a script to switch the java version to be called  at the time `vnc.sh` is.

**Comment by olivergondza** [comment]
I agree with the approach of removing the required build argument. The question is if we want to build separate images for individual java versions we use or make it switchable as implemented here. I do not feel strongly either way.

File: ath-container.sh (hunk 2 above)
Comment (olivergondza) [reply ath-container.sh]: Oh, it is actually never run. Nevermind.

File: set-java.sh [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+wsimport
+wsimport.1.gz
+xjc
+xjc.1.gz"
+
+# Set to the right java version (if the program doesn't exist in this java version, it's not changed)
+for program in $programs
+do
+    # Change the links without showing the choices nor the programs not found
+    echo $selection | update-alternatives --config $program > /dev/null 2> /dev/null
```
Comment (olivergondza) [inline set-java.sh]: I believe [`update-java-alternatives`](https://www.systutorials.com/docs/linux/man/8-update-java-alternatives/) is here to ease this pain.

Comment (MRamonLeon) [reply set-java.sh]: We need root privileges to use _update-java-alternatives_, but we are logging as *ath-user*. I didn't want to change it so I ended up with this approach.

Comment (MRamonLeon) [reply set-java.sh]: We need root privileges to use update-java-alternatives, but we are logging as ath-user. I didn't want to change it so I ended up with this approach.

**Comment by MRamonLeon** [comment]
@oleg-nenashev Do you still consider necessary to add the jaxb modules? According to Oliver's answer, it's not necessary, IIUC.

@olivergondza The instruction (`docker run ...`) changes the JDK and leave you in the shell, ready to do exactly the same as you had to do in previous versions. Do you still request the change? I don't think so.

**Comment by oleg-nenashev** [comment]
> @oleg-nenashev Do you still consider necessary to add the jaxb modules? According to Oliver's answer, it's not necessary, IIUC.

Yes, I do. If it is not necessary, it most likely means there is no tests reproducing the issue. We excluded several usages from the core, so now it does not just crash on the startup.

I am OK if it is excluded from this PR and if there is a follow-up JIRA task for that

**Comment by olivergondza** [comment]
> @olivergondza The instruction (docker run ...) changes the JDK and leave you in the shell, ready to do exactly the same as you had to do in previous versions. Do you still request the change? I don't think so.

Omission on my side. Marking the thread as resolved.

File: set-java.sh (hunk 3 above)
Comment (olivergondza) [reply set-java.sh]: I see. How about giving this file a suid bit for root and use `update-java-alternatives` in it? We would not have to switch files one by one nor changing ownership of `/etc/alternatives`. WDYT?

File: pom.xml [hunk 4]
```diff
@@ -488,7 +488,7 @@
             </property>
           </properties>
           <!-- SUREFIRE-1588 workaround; too late for systemProperties: -->
-          <argLine>-Xmx768M -Djava.awt.headless=true -Djdk.net.URLClassPath.disableClassPathURLCheck=true</argLine>
```
Comment (oleg-nenashev) [inline pom.xml]: how is it related to the change?

File: set-java.sh (hunk 3 above)
Comment (MRamonLeon) [reply set-java.sh]: I tried, but when ´update-java-alternatives´ is executed, I receive: 
_update-java-alternatives: no root privileges_

**Comment by MRamonLeon** [comment]
@oleg-nenashev : Added libraries and modules for java 11
@olivierdagenais : Tested the uid bit without success

File: pom.xml (hunk 4 above)
Comment (MRamonLeon) [reply pom.xml]: are you reviewing commit by commit? :open_mouth: 

I was trying to avoid the _NoClassFoundException_ that I'm still getting from time to time (no idea how I got rid of it and now it's here again):

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test (default-test) on project acceptance-test-harness: Exception in provider: artifactory_plugin/quickstart/bin/src/test/java/org/gradle/PersonTest (wrong name: src/test/java/org/gradle/PersonTest) -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test (default-test) on project acceptance-test-harness: Exception in provider
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoFailureException: Exception in provider
	at org.apache.maven.plugin.surefire.SurefirePlugin.assertNoException(SurefirePlugin.java:328)
	at org.apache.maven.plugin.surefire.SurefirePlugin.handleSummary(SurefirePlugin.java:318)
	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:892)
	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:755)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
	... 20 more
Caused by: org.apache.maven.surefire.booter.SurefireExecutionException: Exception in provider
	at org.apache.maven.plugin.surefire.InPluginVMSurefireStarter.runSuitesInProcess(InPluginVMSurefireStarter.java:87)
	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeProvider(AbstractSurefireMojo.java:1010)
	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:862)
	... 23 more
Caused by: java.lang.NoClassDefFoundError: artifactory_plugin/quickstart/bin/src/test/java/org/gradle/PersonTest (wrong name: src/test/java/org/gradle/PersonTest)
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at org.apache.maven.surefire.booter.IsolatedClassLoader.loadClass(IsolatedClassLoader.java:97)
	at org.apache.maven.surefire.util.DefaultScanResult.loadClass(DefaultScanResult.java:131)
	at org.apache.maven.surefire.util.DefaultScanResult.applyFilter(DefaultScanResult.java:96)
	at org.apache.maven.surefire.junit4.JUnit4Provider.scanClassPath(JUnit4Provider.java:311)
	at org.apache.maven.surefire.junit4.JUnit4Provider.setTestsToRun(JUnit4Provider.java:191)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:134)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray2(ReflectionUtils.java:202)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:155)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:81)
	at org.apache.maven.plugin.surefire.InPluginVMSurefireStarter.runSuitesInProcess(InPluginVMSurefireStarter.java:82)
	... 25 more
```

Comment (oleg-nenashev) [reply pom.xml]: > are you reviewing commit by commit?

I was reviewing the diff since the previous review, and I didn't get what is the purpose of this change from the commit history or the change

**Comment by oleg-nenashev** [comment]
@MRamonLeon I have serious doubts that Artifactory Plugin can operate well with Java 11 taking its dependencies and the long history of issues we had with it. My suggestion would be to report a JIRA ticket and to park it for now. The issue is really weird, but probably it does not worth digging into it right now

**Comment by MRamonLeon** [comment]
@olivergondza I used the same approach as the used in the [PCT](https://github.com/jenkinsci/plugin-compat-tester/pull/86), _one docker image to rule them all_ ;-) (for the time being)
